File: pcm/default_template/test/judge.cpp
#include "../codes/header.hpp"

enum JudgeResult {/*{{{*/
    AC = 0,
    WA = 1,
    RE = 2,
    CE = 3,
    TLE = -2,
    MLE = 4,
    QLE = 5,  // Query limit exceeded
    AIE = 6,  // Additional input error
    NOEXP = 98,
    TLENAIVE = 99,
    YET = 100,
};/*}}}*/

template <class T>/*{{{*/
void tell(T x) {
    cout << x << endl;
    fflush(stdout);
}/*}}}*/

bool check_AIE(){/*{{{*/
    ll tmp;cin>>tmp;
    if (cin.eof()==0) exit(AIE);  // check needles user-output does not exist
}/*}}}*/

int judge_case(ll x) {
    ll num = 22;  // TODO: update query number limit
    while(num--){
        string query;
        cin >> query;
        if (query=="?"){ // query
            ll solver_ans;cin>>solver_ans;
            tell(gcd(solver_ans, x));
        }
        if (query=="!"){ // ans
            ll solver_ans;cin>>solver_ans;
            if (abs(solver_ans-x)<=7 || (solver_ans<=2*x && x<=2*solver_ans)){
                return 0;
            }
            else{
                cerr << "WA: ";
                dump(x, solver_ans);
                exit(WA);
            }
        }
    }
    exit(QLE);
}

int main() {
    // input case
    int Q;
    cin >> Q;
    // Q = 1; if single case
    vector<ll> x(Q);
    rep(q, Q){ cin>>x[q]; }

    // judge
    tell(Q);  // off if single case
    rep(q, Q) {
        cerr << "start judge for: " << x[q] << " [judge]" << endl;;
        judge_case(x[q]); 
    }

    // check_AIE(); // これは何故か動かない。無限待機になってしまう。
    exit(AC);
    return 0;
}

File: pcm/default_template/test/judge.py
import sys
from math import gcd
from pcm.codefile import JudgeResult as jr


def tell(x):
    print(x, flush=True)


def dump(x):
    print(x, ': by judge', file=sys.stderr)


def check_AIE():
    try:
        input()
    except EOFError:
        pass
    else:
        exit(jr.AIE)


def judge_case(x):
    dump(f'start judge for {x=}')
    try:
        num = 22
        while (num):
            num -= 1
            query = list(input().split())
            v = int(query[1])
            if (query[0] == "?"):
                tell(gcd(v, x))

            if (query[0] == "!"):
                if (abs(v - x) <= 7 or (v <= 2 * x and x <= 2 * v)):
                    dump(f"AC: {x=}, {v=}")
                    return 0
                else:
                    dump(f"WA: {x=}, {v=}")
                    sys.exit(jr.WA)
        sys.exit(jr.QLE)
    except Exception as e:
        print(e, file=sys.stderr)
        sys.exit(jr.RE)


def main():  # multi case type
    # input case from a case file
    # case fileの入力をとりきらないとsolver-codeからの出力を入力として受け取れないことに注意。
    Q = int(input())
    x = []
    for i in range(Q):
        x.append(int(input()))

    # judge
    tell(Q)  # TODO: needed only for multi-test-case
    for i in range(Q):
        print('', file=sys.stderr)
        judge_case(x[i])

    # これは何故か動かない。
    # check_AIE()
    sys.exit(jr.AC)


# def main():  # single case type
#     x = int(input())
#     judge_case(x)
#     sys.exit(jr.AC)

if __name__ == "__main__":
    # main()
    print("hello")
